Metadata and runtime support for checking isolated conformances at runtime

DougGregor opened the pull request:
Extend the metadata representation of protocol conformance descriptors to include information about the global actor to which the conformance is isolated (when there is one), as well as the conformance of that type to the GlobalActor protocol. Emit this metadata whenever a conformance is isolated.

When performing a conforms-to-protocol check at runtime, check whether the conformance that was found is isolated. If so, extract the serial executor for the global actor and check whether we are running on that executor. If not, the conformance fails.

In its current state, this implementation is missing some pieces:
- [x] We aren't checking whether any of the conformances used to satisfy conditional requirements are isolated, so we allow isolated conformances to slip through to the wrong isolation domain through that mechanism.
- [ ] We aren't correctly substituting instantiation arguments from the witness table into the global actor type.
- [x] The use of `dlsym` is ugly but probably unavoidable, since we need to downcall into the concurrency library. We'll need a separate implementation for Windows.
- [ ] There is no way to express that a conformance is isolated to a `MainActor`-conforming generic parameter or associated type. The language doesn't handle this right now either, but the runtime should be more general.

DougGregor committed:
Metadata and runtime support for checking isolated conformances at runtime

Extend the metadata representation of protocol conformance descriptors
to include information about the global actor to which the conformance is
isolated (when there is one), as well as the conformance of that type to
the GlobalActor protocol. Emit this metadata whenever a conformance is
isolated.

When performing a conforms-to-protocol check at runtime, check whether
the conformance that was found is isolated. If so, extract the serial
executor for the global actor and check whether we are running on that
executor. If not, the conformance fails.

DougGregor commented:
@swift-ci please test

DougGregor commented on stdlib/public/Concurrency/Task.swift:
```diff
@@ -1518,6 +1518,15 @@ func _taskIsCurrentExecutor(_ executor: Builtin.Executor) -> Bool
+  @usableFromInline
+  internal static func _taskIsCurrentGlobalActor() -> Bool {
+    let executor = unsafe sharedUnownedExecutor
+    return unsafe _taskIsCurrentExecutor(executor: executor.executor, flags: 0)
```
None of the code that would call this entry point back-deploys, so I think it's safe from crash-if-not-expected. If we did try to back-deploy, yeah, it would be a problem.

ktoso replied:
Main executor has a bunch of special case sugar so it'll be fine; but this may come up for global actors on custom executors... there's probably pretty few of them so we may choose to not worry about this too much and just "if it blows up, update your impl" which is probably good enough, just highlighting that it's an issue that'll happen eventually

al45tair commented on stdlib/public/runtime/ProtocolConformance.cpp:
```diff
@@ -1154,6 +1154,50 @@ swift_conformsToProtocolMaybeInstantiateSuperclasses(
+  // The concurrency library provides a function to check whether we
+  // are executing on the given global actor.
+  auto isCurrentGlobalActor = SWIFT_LAZY_CONSTANT(reinterpret_cast<bool (*)(const Metadata *, const WitnessTable *)>(
+      dlsym(RTLD_DEFAULT, "swift_task_isCurrentGlobalActor")));
```
Could we avoid this by having the concurrency library register a function with libswiftCore? `dlsym` is problematic because it won't work on Linux when we're statically linked (and it doesn't exist on Windows either).

mikeash replied:
We already do this sort of thing for Concurrency types with standard mangling. Look at the uses of `ConcurrencyStandardTypeDescriptors`. We could extend that facility to a more general "register stuff from the Concurrency runtime" thing, or add a second one.

DougGregor replied:
That's a great idea. This is runtime SPI, so I'll generalize it rather than adding a second one. Fewer `__attribute__((constructor))`s is a good thing

DougGregor committed:
Test case fixes for isolated conformances

DougGregor committed:
[Isolated conformances] Check witness tables for conditional requirements

When establishing whether a given conformance is isolated, look through
the witness tables used to satisfy conditional requirements as well. This
is because an otherwise-nonisolated conditional conformance can become
isolated if one of its associated conformance requirements is satisfied
by an isolated conformance.

While here, make sure this code works with variadic generics, too.

DougGregor commented:
@swift-ci please test

DougGregor committed:
Replace dlsym of swift_task_isCurrentGlobalActor with a constructor hook

Following the approach taken with the concurrency-specific type
descriptors, register a hook function for the "is current global actor"
check used for isolated conformances.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test Windows

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
First time I broke wasm, yay?

DougGregor committed:
[Isolated conformances] Cache resolved global actor for conformances

In the prior implementation of runtime resolution of isolated conformances,
the runtime had to look in both the protocol conformance descriptor and
in all conditional conformance requirements (recursively) to find any
isolated conformances. If it found one, it had to demangle the global
actor type to metadata. Since swift_conformsToProtocol is a hot path through
the runtime, we can't afford this non-constant-time work in the common
case.

Instead, cache the resolved global actor and witness table as part of the
conformance cache, so that we have access to this information every time
we look up a witness table for a conformance. Propagate this up through
various callers (e.g., generic requirement checking) to the point where
we either stash it in the cache or check it at runtime. This gets us down
to a very quick check (basically, NULL-or-not) for nonisolated conformances,
and just one check for isolated conformances.

DougGregor committed:
Rework runtime entrypoints for isolated conformance checking

Replace the pair of global actor type/conformance we are passing around with
a general "conformance execution context" that could grow new functionality
over time. Add three external symbols to the runtime:

* swift_conformsToProtocolWithExecutionContext: a conforms-to-protocol check
  that also captures the execution context that should be checked before
  using the conformance for anything. The only execution context right now
  is for an isolated conformance.
* swift_isInConformanceExecutionContext: checks whether the function is
  being executed in the given execution context, i.e., running on the
  executor for the given global actor.
* swift_ConformanceExecutionContextSize: the size of the conformance
  execution context. Client code outside of the Swift runtime can allocate
  a pointer-aligned region of memory of this size to use with the runtime
  functions above.

DougGregor committed:
[WebAssembly] Temporarily work around lack of __attribute__((constructor))

We don't have a great way to ensure that the current-global-actor hook
will get installed by the concurrency library with WebAssembly, so
temporarily work around the issue by relying on the fact that we also
aren't doing actual concurrency with WebAssembly.

DougGregor committed:
Generalize FileCheck'ing of IRGen testcase slightly

DougGregor commented:
Based on an idea of @mikeash, I reworked the runtime portion of this to not require walking conditional requirements at lookup time. Instead, we store the isolated-conformance information within the conformance cache itself. The isolation of a conformance is determined at the point when we populate the cache (when we are already checking conditional requirements anyway), and is quick to check when the witness table was already cached. We cache the global actor type + its conformance to GlobalActor, so we also don't require demangle-to-type-metadata after the initiate entry is cached. The only memory over

While here, tidy up the separation between "checks conformance but not the execution context" and "check the execution context" in a manner that can help us with other execution-time checks in the future.

DougGregor commented:
@swift-ci please test

DougGregor committed:
Revert "[WebAssembly] Temporarily work around lack of __attribute__((constructor))"

This reverts commit 8ef9f7fcad73d9d6968bd8e3c7f99e5e56541da5.

DougGregor committed:
[WebAssembly] Disable the runtime test for isolated conformances

WebAssembly doesn't currently have a good way to install the
"isCurrentGlobalActor" hook on which this checking depends. Disable
the test for the moment.

DougGregor committed:
Fixup ABI test

DougGregor commented:
@swift-ci please test

DougGregor committed:
Disable test on WebAssembly harder

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please benchmark

DougGregor committed:
Generalize IRGen test yet more

DougGregor commented:
@swift-ci please test
